src/NumericalAlgorithms/Spectral/SwshCollocation.hpp

```diff
@@ -0,0 +1,245 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <cstdlib>
+#include <memory>
+#include <sharp_cxx.h>
+#include <type_traits>
+#include <utility>
+
+#include "ErrorHandling/Assert.hpp"
+#include "NumericalAlgorithms/Spectral/ComplexDataView.hpp"
+#include "Utilities/ForceInline.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/TMPL.hpp"
+
+// In the static caching mechanism, we permit an l_max up to this macro
+// value. Higher l_max values may still be created manually using the
+// `Collocation` constructor. If l_max's are used several times at higher value,
+// consider increasing this value, but only after memory costs have been
+// evaluated.
+#define SWSH_COLLOCATION_MAXIMUM_L_MAX 200
+
+/// \brief Namespace holding spherical harmonic transformation and helper
+/// utilities for angular data.
+namespace Spectral {
+namespace Swsh {
+
+namespace detail {
+// Helping functor to appropriately delete a stored `sharp_geom_info**`
+struct DestroySharpGeometry {
+  void operator()(sharp_geom_info** to_delete) noexcept {
+    sharp_destroy_geom_info(*to_delete);
+    free(to_delete);
+  }
+};
+}  // namespace detail
+
+/// A container for reporting a single collocation point for libsharp compatible
+/// data structures
+struct LibsharpCollocationPoint {
+  size_t offset;
+  double theta;
+  double phi;
+};
+
+/// \brief A wrapper class for the spherical harmonic library collocation data
+///
+/// \details The currently chosen library for spin-weighted spherical harmonic
+/// transforms is libsharp. The `Collocation` class stores the
+/// libsharp `sharp_geom_info` object, which contains data about
+/// 1. The angular collocation points used in spin-weighted spherical harmonic
+/// transforms
+/// 2. The memory representation of double-type values at those collocation
+/// points
+///
+/// \tparam Representation the ComplexRepresentation, either
+/// `ComplexRepresentation::Interleaved` or
+/// `ComplexRepresentation::RealsThenImags` compatible with the generated
+/// `Collocation` - this is necessary because the stored libsharp type contains
+/// memory stride information.
+template <ComplexRepresentation Representation>
+class Collocation {
+ public:
+  class CollocationIterator {
```

Comment:
I like that this is now a const iterator. I think naming it `CollocationConstIterator` might not be a bad idea. The STL does this quite a bit since `const` iterators can be non-const, but the data they "point" to is const